codegen/src/main.zig

```diff
@@ -634,7 +752,465 @@ fn generateMetadataFunction(operation_name: []const u8, state: GenerationState,
+                map_capture,
+            });
+            try writer.writeAll("{\n");
+        }
+
+        try writer.writeAll("try jw.beginObject();\n");
+        try writer.writeAll("{\n");
+
+        // start loop
+        try writer.print("for ({s}) |{s}|", .{ map_value, map_value_capture });
+        try writer.writeAll("{\n");
+        try writer.print("try jw.objectField({s});\n", .{map_capture_key});
+
+        try writeMemberJson(.{
+            .shape_id = map.value,
+            .field_name = "value",
+            .field_value = map_capture_value,
+            .state = state.indent(),
+            .member = value_member,
+        }, writer);
+
+        try writer.writeAll("}\n");
+        // end loop
+
+        try writer.writeAll("}\n");
+        try writer.writeAll("try jw.endObject();\n");
+
+        if (map_is_optional) {
+            try writer.writeAll("} else {\n");
+            try writer.writeAll("try jw.write(null);\n");
+            try writer.writeAll("}\n");
+        }
+    }
+}
+
+fn writeScalarJson(comment: []const u8, params: WriteMemberJsonParams, writer: std.io.AnyWriter) anyerror!void {
+    try writer.print("try jw.write({s}); // {s}\n\n", .{ params.field_value, comment });
+}
+
+fn writeMemberJson(params: WriteMemberJsonParams, writer: std.io.AnyWriter) anyerror!void {
+    const shape_id = params.shape_id;
+    const state = params.state;
+    const shape_info = try shapeInfoForId(shape_id, state.file_state.shapes);
+    const shape = shape_info.shape;
+
+    if (state.getTypeRecurrenceCount(shape_id) > 2) {
+        return;
+    }
+
+    try state.appendToTypeStack(&shape_info);
+    defer state.popFromTypeStack();
+
+    switch (shape) {
+        .structure, .uniontype => try writeStructureJson(params, writer),
+        .list => |l| try writeListJson(l, params, writer),
+        .map => |m| try writeMapJson(m, params, writer),
+        .timestamp => try writeScalarJson("timestamp", params, writer),
+        .string => try writeScalarJson("string", params, writer),
+        .@"enum" => try writeScalarJson("enum", params, writer),
+        .document => try writeScalarJson("document", params, writer),
+        .blob => try writeScalarJson("blob", params, writer),
+        .boolean => try writeScalarJson("bool", params, writer),
+        .float => try writeScalarJson("float", params, writer),
+        .integer => try writeScalarJson("integer", params, writer),
+        .long => try writeScalarJson("long", params, writer),
+        .double => try writeScalarJson("double", params, writer),
+        .bigDecimal => try writeScalarJson("bigDecimal", params, writer),
+        .bigInteger => try writeScalarJson("bigInteger", params, writer),
+        .unit => try writeScalarJson("unit", params, writer),
+        .byte => try writeScalarJson("byte", params, writer),
+        .short => try writeScalarJson("short", params, writer),
+        else => std.debug.panic("Unexpected shape type: {}", .{shape}),
```

Comment:
This looks like it might be comprehensive of all the possibilities.  If that is true, we should drop the else and let it be a compiler error instead of a runtime panic.

Reply:
Service, resource and operation

Reply:
Updated to handle all types explicitly 6e775e2

Don't think we can use a compiler error because `shape` is runtime known.